# Notes about process of calling SNPs for Camelina resequencing libraries

## Overview
* Generate subdirectories for SNP calling
* Set up snps.config files for each library
* Run snpCaller
* Check quality of snpCaller output
* Clean up unneeded temporary files generated by snpCaller

## Generate directories for SNP calling
### Overview
* use original file for choosing `LIB_CODE`s to generate directories
### Code
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling
for i in `cat /global/projectb/scratch/grabowsp/Csativa_reseq/snp_call_1/camsat_reseq_libs_for_SNPcalling_1.txt`; do mkdir $i; done
```

## SNP Calling for `LIB_CODE`s with multiple fastq files
### Overview
* Ran test with Jerry for CAWTG, so have properly setup snps.config file from \
that that can use as template for the other 3 `LIB_CODE`s that have 2 fastq \
files
* Step 1: copy and adjust the snps.config file using `sed`
* Step 2: generate soft-links for the additional prepped fastqs in the prep \
directory entitled `LIB_CODE`
  * For example, make soft linke for `LIB_CODE_1.prepped.fastq.bz2` in the \
`LIB_CODE` directory
* Step 3: Start `snpCaller_GP.py`
### Generate snps.config files for other 3 `LIB_CODE`s
```
for LIB in CAWTH CAWTN CAWTO;
do sed 's/CAWTG/'"$LIB"'/g' /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/CAWTG/snps.config \
> /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LIB/snps.config; done
```
### Generate soft links for additional prepped fastq files
* already did this for CAWTG when doing test with Jerry
```
cd /global/projectb/scratch/grabowsp/Csativa_reseq/snp_call_1
for LIB in CAWTH CAWTN CAWTO;
do ln -s \
/global/projectb/scratch/grabowsp/Csativa_reseq/snp_call_1/$LIB'_1/'$LIB'_1.prepped.fastq.bz2' \
/global/projectb/scratch/grabowsp/Csativa_reseq/snp_call_1/$LIB'/'$LIB'_1.prepped.fastq.bz2'; done
```
### Start snpCaller
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/
for LIB in CAWTH CAWTN CAWTO;
do cd ./$LIB
snpCaller_GP.py
sleep 10s
cd ..;
done
```
### Clean up directory after snpCalling
* commands from Jerry
* testing for CAWTG:
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/CAWTG
rm -f *.sh.* *fastq.gz *.sorting.sh *.merging.sh
rm -f *.RG.ba* *.deDup.ba* *.reAligned.ba*
rm -f *.GRP_?.?.bam *.GRP_??.?.bam *.GRP_??.??.bam *.GRP_?.??.bam *.GRP_?.???.bam *.GRP_??.???.bam
rm -f *.GRP_?.?.sorted.bam *.GRP_??.?.sorted.bam *.GRP_??.??.sorted.bam *.GRP_?.??.sorted.bam *.GRP_?.???.sorted.bam *.GRP_??.???.sorted.bam
```
* For other 3 `LIB_CODE`s:
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/
for LIB in CAWTH CAWTN CAWTO;
do cd ./$LIB
rm -f *.sh.* *fastq.gz *.sorting.sh *.merging.sh
rm -f *.RG.ba* *.deDup.ba* *.reAligned.ba*
rm -f *.GRP_?.?.bam *.GRP_??.?.bam *.GRP_??.??.bam *.GRP_?.??.bam *.GRP_?.???.bam *.GRP_??.???.bam
rm -f *.GRP_?.?.sorted.bam *.GRP_??.?.sorted.bam *.GRP_??.??.sorted.bam *.GRP_?.??.sorted.bam *.GRP_?.???.sorted.bam *.GRP_??.???.sorted.bam
cd ..;
done
```

## SNP Calling for `LIB_CODE`s with 1 single prepped fastq file
* Steps:
  * Generate list of `LIB_CODE`s
  * Generate generic snps.config
  * Copy and adjust snps.config to each directory
  * Make sublists of 20 `LIB_CODE`s
  * Start snp caller on one sublist at a time
### Generate list of `LIB_CODE`s
* will use the `LIB_CODE` file from prepping and take out the `LIB_CODE`s \
with 2+ fastq files
* In R:
```
lc_file <- '/global/projectb/scratch/grabowsp/Csativa_reseq/snp_call_1/camsat_reseq_libs_for_SNPcalling_1.txt'
lcs <- read.table(lc_file, header = F, sep = '\t', stringsAsFactors = F)
multi_lc <- c('CAWTG', 'CAWTH', 'CAWTN', 'CAWTO')
sing_lc <- setdiff(lcs[,1], multi_lc)

sing_lc_file <- '/global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/single_fastq_libCodes.txt'
write.table(sing_lc, file = sing_lc_file, quote = F, sep = '\t', 
  row.names = F, col.names = F)
```
### Generate generic single-fastq snps.config file
* copy snps.config from CAWTG and make adjustments to use as template for \
single fastq `LIB_CODE`s
### Copy and adjust generic snps.config to each sub-directory
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/
for LIB in `cat single_fastq_libCodes.txt`;
do
sed 's/TEST/'"$LIB"'/g' /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/snps.config_1file > /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LIB'/snps.config';
done
```
### Generate sub-lists
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/
split -l 20 -d single_fastq_libCodes.txt sing_lc_sub_
```
### Start SNPcaller
* ran: `sing_lc_sub_00`; `sing_lc_sub_01`; `sing_lc_sub_02`; `sing_lc_sub_03`;\
`sing_lc_sub_04`; `sing_lc_sub_05`; `sing_lc_sub_06`; `sing_lc_sub_07`; \
`sing_lc_sub_08`; `sing_lc_sub_09`; `sing_lc_sub_10`; `sing_lc_sub_11`
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/
for LIB in `cat sing_lc_sub_11`;
do cd ./$LIB
snpCaller_GP.py
sleep 1s
cd ..;
done
```

## QC
### Overview
* check that final .bam files were generated
### Check that final bam and VCF was generated
* `sing_lc_sub_00`; `sing_lc_sub_01`; `sing_lc_sub_02`; `sing_lc_sub_03`; \
`sing_lc_sub_04`; `sing_lc_sub_05`; `sing_lc_sub_06`
* should check 07, 08, 09, 10, 11 at same time and re-start \
problematic ones for those libraries at the same time
```
bash
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling
for i in `cat single_fastq_libCodes.txt`;
do ls -oh ./$i'/'$i'.GATK.SNP.postFilter.vcf';
done
```

## Troubleshooting
### .bam sorting for GRP_20 did not finish
#### Background
* 7 libraries failed at the same spot: the sorting of the final .bam files \
for GRP_20
* Noticed because got "FAILED" message from SLURM in my email
#### Test Solution for GCNPC
* Jerry recommended trying the following
```
bash
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/GCNPC
for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>0)print $9}' | rev \
| cut -f2- -d"." | rev);
do echo $x; sbatch $x; done
```
* once that is done, try this:
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/GCNPC
snpCaller_GP.py -s merge_bam_files
```
* Worked!
#### Run for remaining 6 libraries
```
for LC in GBSYY GBSXH GCNPU GCNPO GCNPB GCNPS;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>0)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* Next step:
```
for LC in GBSYY GBSXH GCNPU GCNPO GCNPB GCNPS;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  done
```
#### GCNPB didn't finish making vcf files
* trying this...
```
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/GCNPB
snpCaller_GP.py -s GC_Depth_plots
```
### Issue with `sing_lc_sub_02` libraries
* 6 libraries had issues:
  * GBPTZ; GBSZG; GBPTX; GGNUU; GGNWW; GBPUN
#### Try restarting the sorting part
```
bash
for LC in GBPTZ GBSZG GBPTX GGNUU GGNWW GBPUN;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>0)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* Next step:
```
for LC in GBPTZ GBSZG GBPTX GGNUU GGNWW GBPUN;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  done
```
### Issues with `sing_lc_sub_03` libraries
* Several libraries did not finish:
  * GBPSY, GBPXY, GBPUC, GGNWU, GBPZX, GBPWC
* Based on the SLURM error outputs, it looks like the failures occured \
at different points:
  * GBPWC looks to have failed during the merging (like most of the previous \
failures)
  * The others failed during the GC vs depth step, but looking at their \
sorting.sh.stderr outputs, it looks like they failed at same point, so will \
try re-starting them all at that same spot
#### Try restarting the sorting part
* note: it looks like any .sorting.sh.stderr file larger than 122 is a \
failed run, so will change the `awk` part of the command to reflect that
```
bash
for LC in GBPSY GBPXY GBPUC GGNWU GBPZX GBPWC;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>122)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* Next step:
```
for LC in GBPSY GBPXY GBPUC GGNWU GBPZX GBPWC;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  done
```
### Issues with `sing_lc_sub_04` libraries
* Several libraries did not finish:
  * GBPXN; GCNHT; GCNHX; GCNHN; GCNHU
#### Try restarting the sorting part
```
bash
for LC in GBPXN GCNHT GCNHX GCNHN GCNHU;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>122)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* Next step:
```
for LC in GBPXN GCNHT GCNHX GCNHN GCNHU;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  done
```
### Issues with `sing_lc_sub_05` libraries
* GCNHZ; GCNHO; GCNON; GCNCB; GBSXN
```
bash
for LC in GCNHZ GCNHO GCNON GCNCB GBSXN;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>122)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* next step:
```
for LC in GCNHZ GCNHO GCNON GCNCB GBSXN;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  done
```
### Issues with `sing_lc_sub_06` libraries
* GBSYA; GGNWA
```
bash
for LC in GBSYA GGNWA;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>122)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* next step:
```
for LC in GBSYA GGNWA;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  done
```
### Issues with `sub_07`, `sub_08`, `sub_09`, `sub_10`, `sub_11`
* sub_07: GGNXW
  * - looks like "normal" failure
* all the sub_08 libraries completed
* sub_09: GCNCW, GCNCH, GBPYY, GBPYX, GGNWB, GGNUY
  * "normal" failure: GCNCW, GBPYY, GBPYX, GGNWB, GGNUY
  * a lot of failures: GCNCH
    * still looks like can start from the same spot, though
* sub_10: GCNOU, GCNOS, GCNOB, GGNXU, GBSUX, GBPYO
* sub_11: GBPZP, GBPXX
* ones that worked with 1st restart: GBSUX, GBPXX
* the rest worked after the 2nd restart
```
for LC in GGNXW GCNCW GCNCH GBPYY GBPYX GGNWB GGNUY GCNOU GCNOS GCNOB \
GGNXU GBPYO GBPZP;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  for x in $(ls -l *.sorting.sh.stderr | awk '{if($5>122)print $9}' | rev \
| cut -f2- -d"." | rev);
    do echo $x; sbatch $x; sleep 1s;
    done;
  done
```
* next step:
```
for LC in GGNXW GCNCW GCNCH GBPYY GBPYX GGNWB GGNUY GCNOU GCNOS GCNOB \
GGNXU GBPYO GBPZP;
  do cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/$LC;
  snpCaller_GP.py -s merge_bam_files;
  sleep 1s;
  done
```

## Clean up libraries after finishing with SNP calling
* done for `sing_lc_sub_00`, `sing_lc_sub_01`, `sing_lc_sub_02`, \
`sing_lc_sub_03`; `sing_lc_sub_04`; `sing_lc_sub_05`; `sing_lc_sub_06`, \
`sing_lc_sub_07`; `sing_lc_sub_08`; `sing_lc_sub_09`; `sing_lc_sub_10`; \
`sing_lc_sub_11`
```
bash
cd /global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/
for LIB in `cat sing_lc_sub_11`;
do cd ./$LIB
rm -f *.sh.* *fastq.gz *.sorting.sh *.merging.sh
rm -f *.RG.ba* *.deDup.ba* *.reAligned.ba*
rm -f *.GRP_?.?.bam *.GRP_??.?.bam *.GRP_??.??.bam *.GRP_?.??.bam *.GRP_?.???.bam *.GRP_??.???.bam
rm -f *.GRP_?.?.sorted.bam *.GRP_??.?.sorted.bam *.GRP_??.??.sorted.bam *.GRP_?.??.sorted.bam *.GRP_?.???.sorted.bam *.GRP_??.???.sorted.bam
cd ..;
done
```

## Use VarScan to call SNPs
### Make list of bam files to use with VarScan
```
div_dir <- '/global/projectb/scratch/grabowsp/Cs_reseq_snp_calling/'

div_lib_file <- paste('/global/projectb/scratch/grabowsp/Csativa_reseq/', 
  'snp_call_1/camsat_reseq_libs_for_SNPcalling_1.txt', sep = '')

div_libs <- read.table(div_lib_file, header = T, sep = '\t', 
  stringsAsFactors = F)

div_full_path <- paste(div_dir, div_libs[,1], '/', div_libs[,1], '.gatk.bam', 
  sep = '')

div_bam_list_out <- paste('/global/projectb/scratch/grabowsp/', 
  'Cs_reseq_snp_calling/Cs_reseq_varscan/diversity_bam_list.txt', sep = '')

write.table(div_full_path, div_bam_list_out, quote = F, sep = '\t', 
  row.names = F, col.names = F)
```
## Location of VCF generated by Sujan
### Sujans's location
* `/home/f1p1/tmp/sujan/Camelina/Diversity/Camelina_Diversity_231g.vcf.bz2`
### My location
* `/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity`
### Make into gzip version for VCFtools
```
cd /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/
bzip2 -dk Camelina_Diversity_231g.vcf.bz2

head -1000 Camelina_Diversity_231g.vcf > Camsat_div_231_1kSNPS.vcf

pigz Camelina_Diversity_231g.vcf
```

## Filtering of SNPs
### Temporary Filtering Output Directory
* `/home/grabowsky_scratch/Camsat_diversity/snp_filtering`
### Filtering criteria
* Remove SNPs with >20% missing data
  * --max-missing 0.8
* Need 2 alleles
  * --min-alleles 2 --max-alleles 2
* Max heterozygosity 0.7
### Test with short file
```
bash
source activate NGS_analysis
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering

vcftools --vcf \
/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_div_231_1kSNPS.vcf \
--max-missing 0.8 --min-alleles 2 --max-alleles 2 \
--hardy --out Cs1ktest

# --recode --recode-INFO-all --out Cs_1ktest
```
* need to do both separately

## Filter by missing data and number of alleles
### Generate VCF
* `/home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_div_filt1_vcf.sh`

```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
qsub Camsat_div_filt1_vcf.sh
```

## Generate File with counts of each genotype class
* `/home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_div_1_hardy.sh`

```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
qsub Camsat_div_1_hardy.sh
```


## Generate List of SNPs to remove
* Extract the first three lines of the HWE output
  * `cut /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filt2.hwe -f 1-3 > geno_count.txt`
* Make file with SNPs to exlude
  * SNPs with more than 70% heterozygous genotypes
```
in_file <- '/home/grabowsky_scratch/Camsat_diversity/snp_filtering/geno_count.txt'
count_df <- read.table(in_file, header = T, stringsAsFactors = F, sep = '\t')

per_het_cut <- 0.7

count_list <- strsplit(count_df[,3], split = '/')
count_df$n_het <- unlist(lapply(count_list, function(x) as.numeric(x[2])))

n_samp_vec <- unlist(lapply(count_list[1:100], function(x) sum(as.numeric(x))))
tot_n_samp <- median(n_samp_vec)

n_cutoff <- ceiling(tot_n_samp * per_het_cut)

high_inds <- which(count_df$n_het > n_cutoff)

remove_df <- count_df[high_inds, ]

out_file <- '/home/grabowsky_scratch/Camsat_diversity/snp_filtering/bad_snps_highH.txt'
write.table(remove_df[, c(1:2)], file = out_file, quote = F, sep = '\t', 
  row.names = F, col.names = F)
```

## Filter by excess of heterozygous genotypes
* `/home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_div_2_vcf.sh`
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
qsub Camsat_div_2_vcf.sh
```
* End with 4,391,757 SNPs

## Extract stats
### missingness
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
vcftools --vcf Camsat_diversity_filt3.recode.vcf \
--missing-indv --out Camsat_diversity_filt
```
### heterozygosity
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
vcftools --vcf Camsat_diversity_filt3.recode.vcf \
--het --out Camsat_diversity_filt_
```

## Make zipped version of filtered VCF
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
pigz -c Camsat_diversity_filt3.recode.vcf \
> Camsat_diversity_filtered_SNPs.vcf.gz
```

## Copy files to work directory
```
rsync /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filtered_SNPs.vcf.gz /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/

cp /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filt.imiss /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/

cp /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filt_.het /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/
```

## Make List of Sample Names and Library Names
* location of file I made my copying info from Chaofu's email about \
sequencing finishing:
  * `/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/diversity_library_info.txt`

```
in_file <- '/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/diversity_library_info.txt'
lib_info <- read.table(in_file, header = F, sep = '\t', stringsAsFactors = F)

test_split <- unlist(strsplit(lib_info[134,1], split = '\n'))

info_vec <- c(lib_info[c(1:133), 1], test_split)

info_vec_2 <- gsub('    ', '\t', info_vec)

lib_vec <- unlist(lapply(strsplit(info_vec_2, split = '\t'), function(x)
  x[3]))

name_vec <- unlist(lapply(strsplit(info_vec_2, split = '\t'), function(x)
  x[4]))

name_vec_2 <- gsub(' Resequencing', '', name_vec)

spec_vec <- unlist(lapply(strsplit(name_vec_2, split = ' '), function(x) 
  paste(x[1], x[2], sep = ' ')))

name_vec_3 <- gsub('Camelina sativa |Camelina pilosa ', '', name_vec_2)

tot_info_df <- data.frame(lib = lib_vec, sample = name_vec_3, 
  species = spec_vec, stringsAsFactors = F)

# Check that all the library names in the VCF are there
vcf_lib_file <- '/home/grabowsky_scratch/Camsat_diversity/snp_filtering/libs_in_vcf.txt'

vcf_libs <- unlist(read.table(vcf_lib_file, header = F, sep = '\t', 
  stringsAsFactors = F)[1,])

setdiff(tot_info_df$lib, vcf_libs)
#  [1] "GBSUO" "GBSYN" "GBSYO" "GBSYP" "GBSYS" "GBSYT" "GBSYU" "GBSYW" "GBSYX"
# [10] "GCNCU"
## Not sure why these didn't make it
## need to take out of the dataframe

keep_inds <- which(tot_info_df$lib %in% vcf_libs)

tot_info_df_2 <- tot_info_df[keep_inds, ]

# setdiff(vcf_libs, tot_info_df$lib)
[1] "GBSYY" "GBSYZ" "CAWTG" "CAWTH" "CAWTN" "CAWTO"
## need to add info for these

add_lib_vec <- c('GBSYY', 'GBSYZ', 'CAWTG', 'CAWTH','CAWTN', 'CAWTO')
add_samp_vec <- c('MT5', 'MT102', 'MT5', 'MT102', 'Calena', 'CSA036')

small_df <- data.frame(lib = add_lib_vec, sample = add_samp_vec, 
  species = 'Camelina sativa', stringsAsFactors = F)

all_info_df <- rbind(tot_info_df_2, small_df)

out_file <- '/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/diversity_library_to_sample_info.txt'

write.table(all_info_df, file = out_file, quote = F, sep = '\t', 
  row.names = F, col.names = T)
```

## Files that should be sent to collaborators
* `/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filtered_SNPs.vcf.gz`
* `/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/diversity_library_to_sample_info.txt`
* `/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filt.imiss`
* `/home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filt_.het`

## Generate links for files
### Directory that collaborators can use to access data:
* `/home/www/restricted_access/camelina_diversity`
* Note: Only link files here - don't actually move the files
### Make the links
```
ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filtered_SNPs.vcf.gz /home/www/restricted_access/camelina_diversity/Camsat_diversity_filtered_SNPs.vcf.gz

ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/diversity_library_to_sample_info.txt /home/www/restricted_access/camelina_diversity/diversity_library_to_sample_info.txt

ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filt.imiss /home/www/restricted_access/camelina_diversity/Camsat_diversity_filt.imiss

ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filt_.het /home/www/restricted_access/camelina_diversity/Camsat_diversity_filt_.het

```

## Make a more Filtered VCF for Huang
* Goal: 100k SNPs
* Remove SNPs with >10% missing data
  * --max-missing 0.8
* Need 2 alleles
  * --min-alleles 2 --max-alleles 2
* Max heterozygosity 0.5
* Filter by MAF - 0.05?
* Filter by proximity
### First round of filtering with vcfTools
* Missing data
  * --max_missing 0.9
* 2 alleles
  * --min-alleles 2 --max-alleles 2
* MAF 0.05
  * --maf 0.05
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
cp Camsat_div_filt1_vcf.sh Camsat_div_filt1_vcf_v2.sh
```
* manually adjust Camsat_div_filt1_vcf_v2.sh with vim
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
qsub Camsat_div_filt1_vcf_v2.sh
```
* Output is 1,904,039 SNPs

### Generate File with counts of each genotype class
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
cp Camsat_div_1_hardy.sh Camsat_div_1_hardy_v2.sh
qsub Camsat_div_1_hardy_v2.sh
```
* `Camsat_div_1_hardy_v2.sh` manually adjusted using vim

### Generate List of SNPs to remove
#### Extract the first three lines of the HWE output
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering/
cut /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filt_v2.1.hwe -f 1-3 > geno_count_v2.txt
```
#### Make list in R
```
in_file <- paste('/home/grabowsky_scratch/Camsat_diversity/snp_filtering/', 
  'geno_count_v2.txt', sep = '')
count_df <- read.table(in_file, header = T, stringsAsFactors = F, sep = '\t')

per_het_cut <- 0.5

count_list <- strsplit(count_df[,3], split = '/')
count_df$n_het <- unlist(lapply(count_list, function(x) as.numeric(x[2])))

n_samp_vec <- unlist(lapply(count_list[1:100], function(x) sum(as.numeric(x))))
tot_n_samp <- median(n_samp_vec)

n_cutoff <- ceiling(tot_n_samp * per_het_cut)

high_inds <- which(count_df$n_het > n_cutoff)

remove_df <- count_df[high_inds, ]

out_file <- paste('/home/grabowsky_scratch/Camsat_diversity/snp_filtering/', 
  'bad_snps_highH_v2.txt', sep = '')
write.table(remove_df[, c(1:2)], file = out_file, quote = F, sep = '\t',
  row.names = F, col.names = F)

```
* remove 254,785 SNPs

### Filter by excess of heterozygous genotypes
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
cp Camsat_div_2_vcf.sh Camsat_div_hwfilt_vcf_v2.sh
qsub Camsat_div_hwfilt_vcf_v2.sh
```
* Leaves 1,649,254 SNPs

### Prune/thin the SNPs by LD
* Use plink
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering

/home/t4c1/WORK/sujan/softwares/plink2/plink --vcf Camsat_diversity_filt_v2.2.recode.vcf --keep-allele-order --make-bed --allow-extra-chr --out Camsat_div_filt_v2.2

/home/t4c1/WORK/sujan/softwares/plink2/plink --bfile Camsat_div_filt_v2.2 --allow-extra-chr --no-sex --no-parents --indep-pairwise 50 50 0.5 --out Camsat_div_filt_v2.2_ld0.5

/home/t4c1/WORK/sujan/softwares/plink2/plink --bfile Camsat_div_filt_v2.2 --allow-extra-chr --extract Camsat_div_filt_v2.2_ld0.5.prune.in --recode vcf --out Camsat_div_filt_v2.3
```
* Leaves 338,222 SNPs

#### Try using different SNP shifting
```
/home/t4c1/WORK/sujan/softwares/plink2/plink --bfile Camsat_div_filt_v2.2 --allow-extra-chr --no-sex --no-parents --indep-pairwise 50 10 0.4 --out Camsat_div_filt_v2.2_ld0.5_10

/home/t4c1/WORK/sujan/softwares/plink2/plink --bfile Camsat_div_filt_v2.2 --allow-extra-chr --extract Camsat_div_filt_v2.2_ld0.5_10.prune.in --maf 0.1 --recode vcf --out Camsat_div_filt_v2.4
```
* Use this one - it Leaves 165,919 SNPs

### Extract stats
#### missingness
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
vcftools --vcf Camsat_div_filt_v2.4.vcf \
--missing-indv --out Camsat_diversity_filt_v2.4_
```
#### heterozygosity
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
vcftools --vcf Camsat_div_filt_v2.4.vcf \
--het --out Camsat_diversity_filt_v2.4
```

### Make zip of VCF
```
cd /home/grabowsky_scratch/Camsat_diversity/snp_filtering
pigz -c Camsat_div_filt_v2.4.vcf \
> Camsat_diversity_filtered_SNPs_v2.4.vcf.gz
```

### Copy files to work directory
```
rsync /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filtered_SNPs_v2.4.vcf.gz /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/

cp /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filt_v2.4_.imiss /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/

cp /home/grabowsky_scratch/Camsat_diversity/snp_filtering/Camsat_diversity_filt_v2.4.het /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/
```

### Make links in shared directory
```
ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filtered_SNPs_v2.4.vcf.gz /home/www/restricted_access/camelina_diversity/Camsat_diversity_filtered_SNPs_v2.4.vcf.gz

ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filt_v2.4_.imiss /home/www/restricted_access/camelina_diversity/Camsat_diversity_filt_v2.4_.imiss

ln -s /home/t4c1/WORK/grabowsk/data/Camelina_reseq/diversity/Camsat_diversity_filt_v2.4.het /home/www/restricted_access/camelina_diversity/Camsat_diversity_filt_v2.4.het

```
